gh-89480: Document the gdb helpers (#115657)
python/cpython: Doc/howto/gdb_helpers.rst

.. _gdb:

===========
GDB support
===========

.. highlight:: none

If you experience low-level problems such as crashes or deadlocks
(e.g. when debugging extensions or tinkering with parts of CPython
which are written in C), it can be convenient to use a low-level
debugger such as GDB in order to diagnose and fix the issue.
By default, however, GDB (or any of its front-ends) doesn't know about
high-level information specific to the CPython interpreter, such as which
Python function is currently executing, or what type or value has a given
Python object represented by a standard ``PyObject *`` pointer.

> We hereafter present two ways to overcome this limitation.

Reviewer: By looking at the ToC, it's not clear to me what these two ways are.

Reviewer 2: Right, there's only one way left in the document. (The other was for the 2010's.)
Thanks for the catch!

This guide assumes you are familiar with GDB, and focuses on Python-specific
details.

Prerequisites
=============

You need to have:

- GDB 7 or later. (For earlier versions of GDB, see ``Misc/gdbinit`` in the
  sources of Python 3.11 or earlier.)
- GDB-compatible debugging information for Python and any extension you are
  debugging.
- The ``python-gdb.py`` extension.

The extension is built with Python, but might be distributed separately or
not at all. Below, we include tips for a few common systems as examples.
Note that even if the instructions match your system, they might be outdated.

> Python built from source

Reviewer: If this is a prerequisite, it should be listed above, and this section would just explain how to do it, with a title like:
```suggestion
Building Python from source
```

Reviewer 2: I'd keep Python itself is a prerequisite for the prerequisites. The guide doesn't tell you how to get it, but how to set things up for it.
Hopefully “Setup with Python built from source” is clearer.

------------------------

When you build CPython from source, debugging information should be available,
and the build should add a ``python-gdb.py`` file to the root directory of
your repository.

To activate support, you must add the directory containing ``python-gdb.py``
to GDB's "auto-load-safe-path".
If you haven't done this, recent versions of GDB will print out a warning
with instructions on how to do this.

.. note::

>    If you do not seee instructionns for your version of GDB, put this in your

Reviewer: ```suggestion
   If you do not see instructions for your version of GDB, put this in your
```

   configuration file (``~/.gdbinit`` or ``~/.config/gdb/gdbinit``)::

      add-auto-load-safe-path /path/to/cpython

   You can also add multiple paths, separated by ``:``.

> Linux distros

Reviewer: This title is not very descriptive -- maybe
```suggestion
Installing debug information package on Linux
```
If this only works on Linux, you could just use:
```suggestion
Installing Python debug information package
```
and state in the first paragraph that these are only available on Linux.

Reviewer 2: Setting things up depends on how you got your Python. These are examples for two common cases: source builds, and Linux distros. (No Windows or Mac, you wouldn't typically use GDB there.)

-------------

Most Linux systems provide debug information for the system Python
in a package called ``python-debuginfo``, ``python-dbg`` or similar.
For example:

- Fedora:

   .. code-block:: shell

      sudo dnf install gdb
      sudo dnf debuginfo-install python3

- Ubuntu:

   .. code-block:: shell

      sudo apt install gdb python3-dbg

On several recent Linux systems, GDB can download debugging symbols
automatically using *debuginfod*.
However, this will not install the ``python-gdb.py`` extension;
you generally do need to install the debug info package separately.

> Debug build and Development mode

Reviewer: This title could also be more "action-oriented":
```suggestion
Building Python in "debug" and "development" mode
```
or
```suggestion
Enabling "debug" and "development" mode
```
This could also be a sub-section of the "Building Python from source" section above.  You wouldn't want to read that and build everything and then discover that you have to do it again with debug/dev mode enabled.

Reviewer 2: You do want both: for debugging *extensions* it's sometimes easier to use the production build.
And these are not specific to source builds.

================================

For easier debugging, you might want to:

- Use a :ref:`debug build <debug-build>` of Python. (When building from source,
  use ``configure --with-pydebug``. On Linux distros, install and run a package
  like ``python-debug`` or ``python-dbg``, if available.)
- Use the runtime :ref:`development mode <devmode>` (``-X dev``).

Both enable extra assertions and disable some optimizations.
Sometimes this hides the bug you are trying to find, but in most cases they
make the process easier.

> Using the ``python-gdb`` extension

Reviewer: This seems to be the actual how-to.

You could also add a short introductory paragraph like "After installing the dependencies, you can now use the `python-gdb extension to ...".

Reviewer 2: The previous section tells you how to set things up, this tessl you how to use it :)

        #19 (unable to read python frame information)
        #23 (unable to read python frame information)
        #34 (unable to read python frame information)
        #37 Frame 0x9420b04, for file /usr/lib/python2.6/site-packages/gnome_sudoku/main.py, line 906, in start_game ()
            u = UI()
        #40 Frame 0x948e82c, for file /usr/lib/python2.6/site-packages/gnome_sudoku/gnome_sudoku.py, line 22, in start_game (main=<module at remote 0xb771b7f4>)
            main.start_game()

   The frame numbers correspond to those displayed by GDB's standard
   ``backtrace`` command.

``py-print``
------------

   The ``py-print`` command looks up a Python name and tries to print it.
   It looks in locals within the current thread, then globals, then finally
   builtins::

        (gdb) py-print self
        local 'self' = <SwappableArea(running=<gtk.Dialog at remote 0x98faaa4>,
        main_page=0) at remote 0x98fa6e4>
        (gdb) py-print __name__
        global '__name__' = 'gnome_sudoku.dialog_swallower'
        (gdb) py-print len
        builtin 'len' = <built-in function len>
        (gdb) py-print scarlet_pimpernel
        'scarlet_pimpernel' not found

>    If the current C frame corrresponds to multiple Python frames, ``py-print``

Reviewer 3: ```suggestion
   If the current C frame corresponds to multiple Python frames, ``py-print``
```

   only considers the first one.

``py-locals``
-------------

   The ``py-locals`` command looks up all Python locals within the current
   Python frame in the selected thread, and prints their representations::

        (gdb) py-locals
        self = <SwappableArea(running=<gtk.Dialog at remote 0x98faaa4>,
        main_page=0) at remote 0x98fa6e4>
        d = <gtk.Dialog at remote 0x98faaa4>

>    If the current C frame corrresponds to multiple Python frames, locals from

Reviewer 3: ```suggestion
   If the current C frame corresponds to multiple Python frames, locals from
```

Reviewer 2: Arrrr!

Thank you :)

   all of them will be shown::

      (gdb) py-locals
      Locals for recursive_function
      n = 0
      Locals for recursive_function
      n = 1
      Locals for recursive_function
      n = 2
      Locals for recursive_function
      n = 3
      Locals for recursive_function
      n = 4
      Locals for recursive_function
      n = 5
      Locals for <module>

Use with GDB commands

> =====================·

Reviewer: ```suggestion
=====================
```
This broke the formatting of the title -- maybe this is the second way mentioned in the first paragraph?

> The extension comands complement GDB's built-in commands.

Reviewer 3: ```suggestion
The extension commands complement GDB's built-in commands.
```
